sdk/python/kfp/v2/compiler/compiler_utils.py

```diff
@@ -91,3 +93,21 @@ def _get_value(
   return pipeline_spec_pb2.PipelineJob.RuntimeConfig(
       gcs_output_directory=pipeline_root,
       parameters={k: _get_value(v) for k, v in parameter_values.items()})
+
+
+def sanitize_pipeline_name(name: str) -> str:
+  """Sanitizes pipeline name.
+
+  Make the pipeline name conform to [a-z0-9][a-z0-9-]{0,127}.
+  1. Convert upper case to lower case.
+  2. Replace any invalid character with '-'.
+  3. Strip '-' from the left end.
+  4. Truncate the string to at most 128 characters.
+
+  Args:
+    name: The original pipeline name.
+
+  Returns:
+   The sanitized the pipeline name.
+  """
+  return re.sub('[^a-z0-9-]', '-', name.lower()).lstrip('-')[:128]
```

Comment:
sanitizing as a final check seems ok, but it seems preferable here to throw an error if the user's pipeline name does not match the preferred pattern here. I think we want users to be aware of this, and make the necessary change in their pipeline code rather than doing this implicitly for them, especially since this name is later used as an identifier for capturing all runs under the pipeline.

Reply:
The name could come from two sources:
1. User can provide the name via decorator `@dsl.pipeline(name=...)`
2. If no decorator provided (it's optional), then the name is the pipeline function name. In this case, the name has to be transformed as it likely will contain `_` instead of `-`

So I think we have a few options here:
1. Always do the sanitization regardless where the name came from (as this PR currently does).
 Pros: simple implementation, less user effort
 Cons: less user awareness/transparency
2. Throw an error for user provided name in case it fails validation, but auto-sanitize Python function name
 Pros: user awareness
 Cons: inconsistent behavior and complex code logic.
3. Make decorated required, and throw an error in case it doesn't conform to the regex.
 Pros: user awareness and consistent behavior
 Cons: not back-compatible. existing code may not have decorator.

WDYT?

Reply:
I think the dsl.pipeline should not be optional. It would be preferable for us to explicitly ask for the pipeline name, which will translate to the MLMD pipeline context used to group all outputs. I should be able to change the function name but keep the pipeline name constant. This behaviour should be explicit.

Reply:
In that case, we have two routes to achieve the same goal:

- Option 1. Make `dsl.pipeline` decorator required, and throw error on illegal names
  With this option, we will need two checks:
   1. Check if `dsl.pipeline(name=...)` is called (with `name` being a required parameter). This likely requires setting a flag inside the decorator function body.
   1. Check if the name conforms to the regex here.

- Option 2. Throw error on illegal names, but don't explicitly enforce the usage of `dsl.pipeline` decorator. 
  There's only one check: If the name conforms to the regex. But it implicitly requires calling `dsl.pipeline` to set the name, because we currently have a step to convert the pipeline function name into a human readable name:
https://github.com/kubeflow/pipelines/blob/d418f5765472a079b9edf4f13b93801d26f83831/sdk/python/kfp/components/_naming.py#L56-L60
With the `capitalize()` call, the name will never pass the regex check. So we're actually enforcing the decorator, but there's no check for it (and likely we won't document that it's required).

I think we can start with Option 2? Explicitly enforcing `dsl.pipeline` decorator can be a separate story and a standalone PR anyway.

Reply:
I created a new PR for Option 2: #4851 
Will close this one if that gets merged.

Reply:
Starting with option 2 seems fine, but I think option 2 should be the end goal (users should give an "id" to their pipeline).